0000-backtrace-in-core.md

```diff
@@ -0,0 +1,145 @@
+
+```rust
+/// Global implementation of backtrace functionality. Called to create
+/// `RawBacktrace` trait objects.
+#[cfg(not(bootstrap))]
+extern "Rust" {
+    #[lang = "backtrace_create"]
+    fn backtrace_create(ip: usize) -> *mut dyn RawBacktrace;
+
+    #[lang = "backtrace_enabled"]
+    fn backtrace_enabled() -> bool;
+
+    #[lang = "backtrace_status"]
+    fn backtrace_status(raw: *mut dyn RawBacktrace) -> BacktraceStatus;
+}
+
+#[cfg(bootstrap)]
+unsafe fn backtrace_create(_ip: usize) -> *mut dyn RawBacktrace {
+    UnsupportedBacktrace::create().inner
+}
+
+#[cfg(bootstrap)]
+unsafe fn backtrace_enabled() -> bool {
+    false
+}
+
+#[cfg(bootstrap)]
+unsafe fn backtrace_status(_raw: *mut dyn RawBacktrace) -> BacktraceStatus {
+    BacktraceStatus::Unsupported
+}
+
+```
+
+This change will make the `Backtrace` an optional part of the library and 
+This way, if the `Backtrace` is not enabled (on the library level) there is no need for it to report 
+
+Implementation-wise, `Backtrace` is **declared** in the `core` module and **defined** in `std` via *lang_items* which act as function hooks which are resolved during link-time. Special type `StdBacktrace` is introduced for the `std` part of implementation which acts as a proxy for `Debug` and `Display` trait impls.
+
+// TODO: add examples of how would one implement these functions themselves like panic hooks
+
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There are actually many drawbacks, most important of them being `Backtrace` using a lot of `std` for the actual backtrace capturing. 
+
+The other one is a potential code bloat in `no_std` contexts, so a possible alternative may be only enabling the `Backtrace` conditionally via `cfg` settings. (not so sure about this though)
+
+
+`no_std` code bloat - Mara mentioned out
+`Error:backtrace` also seems to be blocking??
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The proposed solution is the one which is currently implementable. However, if the Generic Member Access RFC was implemented (link!) we would not have to move the `Backtrace` to `core` at all. In the alternative solution, we would leave the `Backtrace` as it is and instead the `Error` trait will provide a `backtrace()` method which will use the Generic Access to extract the concrete `Backtrace` out of the propagated error.
+
+Alternatively, since the actual implementation of `Backtrace` uses allocating structures like `Box` for backtrace creation, it would be prudent to move it to where this type resides. This will in turn allow users of `alloc` module in embedded and `no_std` contexts to use this functionality without `std`. _should we go for this though instead of the core when we introduce such big changes?_
```

Comment:
This would mean we could only move Error to alloc too (c.f., core), and I think we wanted Error in core so it can be used there.

Reply:
iirc with generic member access, the `Error` trait wouldn't need to explicitly mention `Backtrace`, so `Error` could still be in `core`

Reply:
Oh I misunderstood, I thought this alternative was for the current proposal but to alloc, rather than generic member access with something in alloc